lib/garbage-deleter.js

```diff
@@ -0,0 +1,893 @@
+                name: METRIC_PREFIX + 'instruction_files_queued_count',
+                help: 'Gauge indicating number of instructions files in queue'
+            })
+        };
+    } else {
+        self.metrics = {};
+    }
+    self.addCounter('deleteErrorCount', 0);
+    self.addCounter('deleteCountTotal', 0);
+    self.addCounter('deleteTimeSeconds', 0);
+    self.addCounter('instructionFilesBad', 0);
+    self.addCounter('instructionFilesDeleted', 0);
+    self.addCounter('instructionFilesProcessed', 0);
+    self.addCounter('instructionLinesProcessed', 0);
+    self.setGauge('instructionFilesQueued', 0);
+
+    // Add properties which should be modified only for testing purposes.
+    self.fsCreateReadStream = opts._fsCreateReadStream || fs.createReadStream;
+    self.fsReaddir = opts._fsReaddir || fs.readdir;
+    self.fsReadFile = opts._fsReadFile || fs.readFile;
+    self.fsRename = opts._fsRename || fs.rename;
+    self.fsUnlink = opts._fsUnlink || fs.unlink;
+    self.fsWatch = opts._fsWatch || fs.watch;
+    if (opts._processFileHook) {
+        self._processFileHook = opts._processFileHook;
+    }
+}
+
+GarbageDeleter.prototype.addCounter = function addCounter(counterName, value) {
+    var self = this;
+
+    // For tests, we don't want to require a full metricManager, so in that case
+    // we just manually manage the values in the "metrics" object.
+    if (!self.metricsManager) {
+        if (!self.metrics.hasOwnProperty(counterName)) {
+            self.metrics[counterName] = 0;
+        }
+        self.metrics[counterName] += value;
+        return;
+    }
+
+    self.metrics[counterName].add(value);
+};
+
+GarbageDeleter.prototype.getCounter = function getCounter(counterName) {
+    var self = this;
+
+    if (!self.metricsManager) {
+        return self.metrics[counterName];
+    }
+
+    return self.metrics[counterName].getValue();
+};
+
+GarbageDeleter.prototype.getGauge = function getGauge(gaugeName) {
+    var self = this;
+
+    if (!self.metricsManager) {
+        return self.metrics[gaugeName];
+    }
+
+    return self.metrics[gaugeName].getValue();
+};
+
+GarbageDeleter.prototype.setGauge = function setGauge(gaugeName, value) {
+    var self = this;
+
+    // For tests, we don't want to require a full metricManager, so in that case
+    // we just manually manage the values in the "metrics" object.
+    if (!self.metricsManager) {
+        self.metrics[gaugeName] = value;
+        return;
+    }
+
+    self.metrics[gaugeName].set(value);
+};
+
+GarbageDeleter.prototype.readLines = function readLines(filename, callback) {
+    var self = this;
+
+    var closing = false;
+    var lineReader;
+    var lines = [];
+    var lineNum = 0;
+    var readStream;
+
+    readStream = self.fsCreateReadStream(filename, {
+        encoding: 'utf8',
+        end: (self.maxLineLength + 1) * self.maxLines,
+        start: 0
+    });
+
+    lineReader = readline.createInterface({
```

Comment:
The reason I did it this way was that I can't use `fs.readFile` and specify a maximum read size. If someone dropped a 2G file in the instruction dir accidentally, we'd blow up on memory in a loop. We *could* try doing a fs.stat first to make sure the file is not too large, but then we also have a race condition where the file could still be being written to between the `fs.stat` and the `fs.readFile` and we can't prevent running out of memory.

I suppose what I *can* do is use `fs.open` and then do an `fs.read` with a length parameter from the resulting file descriptor. I'll give this a try. This will require more working memory but still will be limited to the size we've specified and we're not allowing huge files so that should be ok I think.

Reply:
Understood that `fs.readFile` can't be used.  What about this:

```
var s = fs.createReadStream(filename, {start:0, end: MAX_LENGTH-1});

var chunks = [];
var len = 0
s.on('readable', function() {
    var chunk;
    while (chunk = this.read()) {
        len += Buffer.byteLength(chunk);
        chunks.push(chunk);
    }
});
s.on('close', function onClose() {
    var data = Buffer.concat(chunks, len);
    console.log('read %s: %d bytes', filename, data.length);
    //var stringContent = data.toString('utf8');
});
```

Reply:
Sure, I can change to something like that. Thanks.

Reply:
I changed to use your pattern. Hopefully that's better?

Reply:
> Hopefully that's better?

Do you have an opinion?

Reply:
I think this will work. As long as we don't ever want to allow huge instruction files. If we do, we can revisit.